File: adapters/utility-adapter/nodejs/src/adapter.ts

import { ValidateRequest, ValidateResponse, DispenseRequest, DispenseResponse, TransactionStatusResponse } from './types';

export abstract class EFNUtilityAdapter {
    abstract validateCustomer(req: ValidateRequest): Promise<ValidateResponse>;
    abstract dispenseValue(req: DispenseRequest): Promise<DispenseResponse>;
    abstract transactionStatus(efnReference: string): Promise<TransactionStatusResponse>;
}

File: adapters/utility-adapter/rust/src/adapter.rs
use crate::models::*;
use axum::async_trait;

#[async_trait]
pub trait EFNUtilityAdapter: Send + Sync {
    async fn validate_customer(&self, req: ValidateRequest) -> ValidateResponse;
    async fn dispense_value(&self, req: DispenseRequest) -> DispenseResponse;
    async fn transaction_status(&self, efn_reference: String) -> TransactionStatusResponse;
}
